apps/demos/Demos/Chat/MessageEditing/jQuery/data.js

```diff
@@ -0,0 +1,71 @@
+const getTimestamp = function (date, offsetMinutes = 0) {
+  return date.getTime() + offsetMinutes * 60000;
+};
+
+const date = new Date();
+date.setHours(0, 0, 0, 0);
+
+const currentUser = {
+  id: 'c94c0e76-fb49-4b9b-8f07-9f93ed93b4f3',
+  name: 'John Doe',
+};
+
+const supportAgent = {
+  id: 'd16d1a4c-5c67-4e20-b70e-2991c22747c3',
+  name: 'Support Agent',
+  avatarUrl: '../../../../images/petersmith.png',
+};
+
+const messages = [
+  {
+    id: new DevExpress.data.Guid(),
+    timestamp: getTimestamp(date, -9),
+    author: supportAgent,
+    text: 'Hello, John!\nHow can I assist you today?',
+  },
+  {
+    id: new DevExpress.data.Guid(),
+    timestamp: getTimestamp(date, -7),
+    author: currentUser,
+    text: "Hi, I'm having trouble accessing my account.",
+  },
+  {
+    id: new DevExpress.data.Guid(),
+    timestamp: getTimestamp(date, -7),
+    author: currentUser,
+    text: 'It says my password is incorrect.',
+  },
+  {
+    id: new DevExpress.data.Guid(),
+    timestamp: getTimestamp(date, -7),
+    author: currentUser,
+    isDeleted: true,
+  },
+  {
+    id: new DevExpress.data.Guid(),
+    timestamp: getTimestamp(date, -7),
+    author: supportAgent,
+    text: 'I can help you with that. Can you please confirm your UserID for security purposes?',
+    isEdited: true,
+  },
+];
+
+const editingStrategy = {
+  enabled: true,
+  disabled: false,
+  custom: ({ component, message }) => {
+    const { items, user } = component.option();
+
+    const lastMessage = [...items]
+      .reverse()
+      .find((m) => m.author?.id === user.id);
+
+    return !lastMessage.isDeleted && lastMessage?.id === message.id;
+  },
+};
```

Comment:
1. I don't think it's data, it's a business logic. In Angular it'll be places in component.ts not in service, right?
2. Last suitable message should be always possible to be edited, even after currently last message remove or adding of a new one. Now it does now work if you remove last message. It was previously discussed with PMs - we here demonstrate a real scenario how some messengers work
3. Currently this code takes O(n) time on every bubble click, actually real time is 3N for every click. I don't think i'd like to implement it this way in my app. I suggest you better keep only `lastEditableMessageId ` and update it on init/renderMessage with O(1) and on last message remove with O(n) (rare scenario) [but anyway it's better to go from last and update, to improve perf more]

Reply:
1. right
2. https://github.com/DevExpress/DevExtreme/pull/29780#discussion_r2075728684. I don't remember this decision, so will return initial solution back
3. It may do the logic more difficult. Well, will improve for performance purposes